### PR TITLE
Enhanced Volume Mounting: Multiple Mounts and File Support

### PR DESCRIPTION
**Pull request by Cabecinha84**
---
  Enhanced Volume Mounting for Flux Applications

  Overview

  This PR implements comprehensive enhancements to the Flux application volume mounting system, adding support for multiple independent mounts per component, file mounts, cross-component volume sharing, and
  modernizes the Docker API usage. All changes maintain full backward compatibility with existing applications.

  Key Features

  1. Multiple Mount Types per Component

  Applications can now define multiple mount points with different types using pipe-separated syntax:

  - Primary mount: [flags]:<path> (e.g., r:/data)
  - Additional directories: m:<subdir>:<path> (e.g., m:logs:/var/log)
  - File mounts: f:<filename>:<path> (e.g., f:config.yaml:/etc/config.yaml)
  - Component references: <number>:<path> (e.g., 0:/shared)
  - Component subdirectories: c:<number>:<subdir>:<path> (e.g., c:0:backups:/backups)
  - Component files: cf:<number>:<filename>:<path> (e.g., cf:0:cert.pem:/etc/ssl/cert.pem)

  Example:
  containerData: "r:/data|m:logs:/var/log|f:config.yaml:/etc/app.yaml|0:/shared"

  2. File Mount Support

  Applications can now mount individual files (not just directories). Files are created empty during installation, allowing the application to initialize them on first run while persisting changes across container
  restarts.

  Use Cases:
  - Configuration files (e.g., config.yaml, settings.json)
  - Certificate files (e.g., cert.pem, key.pem)
  - Lock files or state files
  - Any file that needs to persist but should be app-initialized

  3. Cross-Component Volume Sharing

  Components can reference and mount volumes from other components within the same application:

  - Component ordering: Components can only reference lower-indexed components
  - Flexible sharing: Mount entire primary volumes or specific subdirectories/files
  - Validated references: Automatic validation of component indices and ordering

  4. Modern Docker API Migration

  Migrated from legacy Docker Binds API to modern Mounts API:

  Before:
  HostConfig: {
    Binds: ['host:container', ...]
  }

  After:
  HostConfig: {
    Mounts: [{
      Type: 'bind',
      Source: 'host',
      Target: 'container',
      BindOptions: { Propagation: 'rprivate' }
    }]
  }

  Technical Changes

  New Modules

  mountParser.js (394 lines)

  - Parses enhanced containerData syntax into structured mount definitions
  - Comprehensive security validations (path traversal prevention, reserved names, etc.)
  - Support for 6 mount types with extensible architecture
  - Duplicate detection for paths and subdirectories

  volumeConstructor.js (263 lines)

  - Constructs Docker Mount objects from parsed definitions
  - Component reference validation and path construction
  - Host path generation for local and component mounts
  - Restart policy determination based on flags

  mountParser.test.js (278 lines)

  - Comprehensive test suite with 29 passing tests
  - Coverage for all mount types, security validations, and edge cases
  - Backward compatibility verification

  Modified Modules

  advancedWorkflows.js

  - Complete rewrite of app volume creation logic
  - Creates all required directories and files based on parsed mount definitions
  - Simplified Syncthing configuration (one-time setup per appdata folder)
  - Proper file/directory permissions and ownership

  dockerService.js

  - Replaced ~50 lines of manual parsing with clean module calls
  - Migrated to modern Docker Mounts API
  - Enhanced error handling and validation
  - Removed experimental feature warnings

  appUtilities.js

  - Fixed storage calculation bug with nested mounts
  - Prevents double-counting of overlapping volume paths
  - Added error handling for storage calculation failures

  File System Structure

  All additional mounts are organized within the appdata directory:

  /apps/component_appname/
    ├── .stfolder            # If syncthing enabled
    ├── .stignore            # If syncthing enabled
    └── appdata/
        ├── (primary data)
        ├── logs/            # m:logs:/var/log
        ├── cache/           # m:cache:/tmp/cache
        ├── config.yaml      # f:config.yaml:/etc/config.yaml
        └── cert.pem         # f:cert.pem:/etc/ssl/cert.pem

  Security Features

  - Path traversal prevention: Blocks .. in paths and subdirectory names
  - Absolute path enforcement: All container paths must start with /
  - Null byte protection: Prevents null byte injection attacks
  - Reserved name blocking: Prevents use of appdata, ., ..
  - Invalid character filtering: Validates subdirectory and filename characters
  - Duplicate detection: Prevents duplicate container paths and local subdirectories
  - Component ordering: Enforces forward-only references to prevent circular dependencies
  - Path length validation: 255 character limit for subdirectory/filename

  Backward Compatibility

  ✅ All existing containerData syntax continues to work:
  - Simple mounts: /data or r:/data
  - Component references: 0:/shared|1:/config
  - All flag combinations: r:, g:, s:, rgs:

  No migration required for existing applications.

  Breaking Changes

  ⚠️ None - This is a fully backward-compatible enhancement.

  Testing

  - Unit tests: 29 tests covering all mount types and edge cases (100% passing)
  - Test command: env NODE_CONFIG_DIR=$PWD/tests/unit/globalconfig mocha tests/unit/mountParser.test.js --exit
  - Coverage: Parser validation, mount construction, security checks, backward compatibility

  Statistics

  - Files changed: 6
  - New files: 3 (935 lines)
  - Modified files: 3 (250 lines changed)
  - Net addition: ~1,108 lines
  - Tests added: 29 unit tests

  Documentation

  Updated CLAUDE.md with comprehensive documentation including:
  - Complete syntax reference with examples
  - Module descriptions and architecture
  - File system structure explanation
  - Important constraints and limitations
  - Testing instructions

  Migration Path

  For new applications:
  // Use enhanced syntax
  containerData: "r:/data|m:logs:/var/log|f:config.yaml:/etc/config.yaml"

  For existing applications:
  // No changes required - existing syntax still works
  containerData: "r:/data"

  Future Enhancements

  Potential future improvements (not in this PR):
  - Read-only mount support
  - Custom propagation modes per mount
  - Volume mounts (in addition to bind mounts)
  - Mount options (noexec, nosuid, etc.)
  - Configurable file permissions per mount

  Checklist

  - Code follows project style guidelines
  - All tests passing
  - Documentation updated (CLAUDE.md)
  - Backward compatibility maintained
  - Security validations implemented
  - No breaking changes introduced
  - Error handling implemented
  - Logging added for debugging

### CODE REVIEW

**Comment by XK4MiLX** [comment]
During deployment (advancedWorkflows.js:407):
const execFile = `sudo touch ${appsFolder + appId}/appdata/${pathInfo.name}`;
await cmdAsync(execFile);
This only creates an empty file. There is no code that:
❌ Accepts file content in the deployment spec
❌ Uploads files during deployment
❌ Populates the file with initial data
❌ Downloads file content from a URL
❌ Copies from a template

**Comment by Cabecinha84** [comment]
> We can't put 10Mb in the specs:
> 
> Max content size: 10MB (decoded)
> 
> It's actualy 13.33 mb as base64 encoding is 4/3.
> 
> I don't think it makes sense to store this kind of thing on the specs. It should be encrypted on flux drive or something and we just use a link (and decrypt somehow)
> 
> We probably need to have a think about the architecture here and make a decision on what is to be stored on specs and where we draw the line, as we are going to end up with huge spec bloat

agreed, for now have removed that functionality. On new app specs we might re add it, and make user share the link for the file or upload to flux storage.

**Comment by Cabecinha84** [comment]
updated PR description

File: ZelBack/src/services/appLifecycle/advancedWorkflows.js [hunk 1]
```diff
@@ -405,47 +507,43 @@ async function createAppVolume(appSpecifications, appName, isComponent, res) {
...
+        status: '.stfolder created',
+      };
+      log.info(stFolderCreation2);
+      if (res) {
+        res.write(serviceHelper.ensureString(stFolderCreation2));
+        if (res.flush) res.flush();
+      }
+
+      // Create .stignore file to exclude backup directory (in parent directory)
+      const stignore = `sudo echo '/backup' >| ${appsFolder + appId}/.stignore`;
```
Comment (MorningLightMountain713) [inline ZelBack/src/services/appLifecycle/advancedWorkflows.js:536]: I'm not 100% sure on this but shouldn't this be "backup" not "/backup"? as it's a child dir and not root?

**Comment by MorningLightMountain713** [comment]
Can you please also add detailed documentation in the docs directory for volume-mounts. With examples that cover each option, and how to use it.

Also some advanced usage examples.

It is very difficult to understand without examples

Especially some explanations around what directory is relative to what etc and where the source files are coming from for the file mounts

**Comment by Cabecinha84** [comment]
> Does this PR sum the usage across mounts now? When assessing usage per app etc?

Question: Does the PR aggregate usage across multiple mounts when calculating per-app usage metrics?

  Answer: ✅ YES - The code properly aggregates usage and prevents double-counting.

  How it works:
  - The getContainerStorage function in appUtilities.js:197-278 has built-in nested mount detection
  - When multiple mounts exist (primary + subdirectories/files), it identifies nested mounts
  - Nested mounts are skipped to avoid double-counting
  - Only the parent appdata/ directory is measured once using du -sb
  - This provides accurate total storage usage for billing/limits

  Example:
  Primary:  /apps/fluxapp/appdata           → COUNTED (calculates entire dir)
  Additional: /apps/fluxapp/appdata/logs      → SKIPPED (nested, already counted)
  Additional: /apps/fluxapp/appdata/config.yaml → SKIPPED (nested, already counted)

**Review by MorningLightMountain713 — APPROVED** [review]
Ack. Just be good if we could start doing up MDs in the docs dir for users (and us) to hunt down without having to read the code.

Just the one comment about the syncthing backup dir that I wasn't sure about. Think it's suppsed to be `backup` not `/backup`

File: ZelBack/src/services/appLifecycle/advancedWorkflows.js (hunk 1 above)
Comment (Cabecinha84) [reply ZelBack/src/services/appLifecycle/advancedWorkflows.js:536]: going to be honest, haven't tested the backups, but I believe they are now exactly the same way, a folder on ${appsFolder + appId}, so nothing should be changed there

Comment (alihm) [reply ZelBack/src/services/appLifecycle/advancedWorkflows.js:536]: The `>|` operator will be calling with user's permissions, not with sudo. Can use this instead:
```const stignore = `sudo bash -c "echo '/backup' > ${appsFolder + appId}/.stignore"`;```
